### ISSUE TITLE
Cart and Checkout block pages don't display notices

### ISSUE THREAD

**Issue by james-allan** [woothemes/woocommerce#50153]
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

WC notices enqueued via `wc_add_notice()` or similar are not shown on the cart or checkout block pages. I'm not sure if this is intended however it causes some behaviour minor UX issues with plugins like Subscriptions. 

### Expected behavior

Notices are shown on the shortcode cart and checkout pages, so I would have expected them to also be shown on the cart and checkout pages too. 

### Actual behavior

Notices aren't printed out. Oddly they are printed on the cart page (but still not the checkout page) if you use the storefront theme, however that's because the theme manually triggers the notices to be displayed. 

### Steps to reproduce

1. Enable the **"Redirect to the cart page after successful addition"** setting in **WooCommerce > Settings > Products**
2. Enable a default theme like twenty-twenty-four
3. Add a product to the cart from the product page.
4. Notice you are redirected to the block cart page but no notice is shown.

If you use a shortcode cart page, you would see the _"x" has been added to your cart_ type notice.
 
<img width="1316" alt="Screenshot 2024-07-31 at 5 54 22 PM" src="https://github.com/user-attachments/assets/12013023-539e-499c-b5ef-35c1417c80f4">

### WordPress Environment

<details><summary>System Status Report</summary>

```
### WordPress Environment ###

WordPress address (URL): [Redacted]
Site address (URL): [Redacted]
WC Version: 9.1.2
Legacy REST API Package Version: The Legacy REST API plugin is not installed on this site.
Action Scheduler Version: ✔ 3.7.4
Log Directory Writable: ✔
WP Version: 6.6.1
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.43 (Unix)
PHP Version: 8.0.22
PHP Post Max Size: 1,000 MB
PHP Time Limit: 1200
PHP Max Input Vars: 4000
cURL Version: 8.4.0
(SecureTransport) LibreSSL/3.3.6

SUHOSIN Installed: –
MySQL Version: 5.7.28
Max Upload Size: 300 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

[REDACTED]

### Post Type Counts ###

attachment: 28
dummy_post: 2
page: 13
post: 7
product: 180
product_variation: 162
revision: 53
shop_coupon: 8
shop_order: 307
shop_subscription: 167
user_request: 1
wc_membership_plan: 1
wc_user_membership: 2
wp_global_styles: 1
wp_navigation: 1
wp_template: 2

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (1) ###

WooCommerce: by Automattic – 9.1.2 (update to version 9.1.4 is available)

### Inactive Plugins (44) ###

Action Scheduler: by Automattic – 3.6.4
Action Scheduler: by Automattic – 3.7.4
Ajax add to cart for WooCommerce: by QuadLayers – 2.2.9
AutomateWoo: by WooCommerce – 5.7.5 (update to version 6.0.30 is available)
Debug: by James Allan – 1.0.0
Debug: by James Allan – 1.0.0
eCommerce Configuration By WP Engine: by WP Engine – 0.2.11
Jetpack: by Automattic – 12.9.1
Log HTTP Requests: by FacetWP
LLC – 1.4.1

Query Monitor: by John Blackbourn – 3.15.0
Relative URL: by Tunghsiao Liu – 0.1.8
Test wcpay PR 6643: by Automattic –
User Switching: by John Blackbourn & contributors – 1.7.0
WCPay Server Redirect: by Automattic –
WCPay Subscription Generator: by James Allan –
Woo All Products For Subscriptions: by Woo – 4.1.5 (update to version 6.0.1 is available)
WooCommerce Bookings: by WooCommerce – 2.1.3 (update to version 2.1.5 is available)
WooCommerce Deposits: by WooCommerce – 2.1.7 (update to version 2.3.0 is available)
WooCommerce EU VAT Number: by WooCommerce – 2.8.8 (update to version 2.9.6 is available)
WooCommerce Eway Payment Gateway: by WooCommerce – 3.6.1 (update to version 3.7.2 is available)
WooCommerce Legacy REST API: by WooCommerce – 1.0.4
WooCommerce Memberships: by SkyVerge – 1.25.2 (update to version 1.26.7 is available)
WooCommerce Payfast Gateway: by WooCommerce – 1.6.2 (update to version 1.6.6 is available)
WooCommerce Payments Dev Tools: by Automattic –
WooCommerce PayPal Payments: by WooCommerce – 2.8.2
WooCommerce Pre-Orders: by WooCommerce – 2.1.2
WooCommerce Product Add-ons: by WooCommerce – 6.4.7 (update to version 7.0.0 is available)
WooCommerce Purchase Order Payment Gateway: by WooCommerce – 1.4.3 (update to version 1.4.7 is available)
WooCommerce SnapScan Gateway: by WooCommerce – 1.2.3
WooCommerce Square: by WooCommerce – 4.7.0 (update to version 4.7.2 is available)
WooCommerce Stripe Gateway: by WooCommerce – 7.9.1
WooCommerce Stripe Gateway: by WooCommerce – 8.5.2
WooCommerce Subscription Downloads: by WooCommerce – 1.4.0 (update to version 1.4.1 is available)
WooCommerce Subscriptions: by WooCommerce – 6.5.0
WooCommerce Subscriptions - Reset Next Payment Dates: by Prospress Inc. – 1.1.0
WooCommerce Subscriptions Core: by Automattic – 7.2.0
WooCommerce Subscriptions CSV Importer and Exporter: by Prospress Inc – 2.1.0
WooCommerce Subscriptions Gifting: by WooCommerce – 2.7.0
WooPayments: by WooCommerce – 7.8.1 (update to version 7.9.2 is available)
WP Console: by Edi Amin – 2.4.1
WP Crontrol: by John Blackbourn & crontributors – 1.16.1
WP Job Manager: by Automattic – 1.42.0
WP Job Manager - WooCommerce Paid Listings: by Automattic – 2.9.8
WP phpMyAdmin: by Puvox.software – 5.2.1.02

### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
job_package (job_package)
job_package_subscription (job_package_subscription)
new-variable (new-variable)
resume_package (resume_package)
resume_package_subscription (resume_package_subscription)
simple (simple)
subscription (subscription)
variable (variable)
variable subscription (variable-subscription)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: ✔
Enforce Approved Product Download Directories: ✔
HPOS feature enabled: ✔
Order datastore: Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore
HPOS data sync enabled: ✔

### Logging ###

Enabled: ✔
Handler: Automattic\WooCommerce\Internal\Admin\Logging\LogHandlerFileV2
Retention period: 30 days
Level threshold: –
Log directory size: 16 MB

### WC Pages ###

Shop base: #6 - /shop/
Cart: #430 - /block-cart/ -  Contains the woocommerce/cart block
Checkout: #187 - /block-checkout/ -  Contains the woocommerce/checkout block
My account: #9 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Twenty-Four
Version: 1.2
Author URL: https://wordpress.org
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ❌ Not declared

### Templates ###

Overrides: /Users/jamesallan/local-dev/wcdev/app/public/wp-content/plugins/woocommerce/templates/block-notices/error.php
/Users/jamesallan/local-dev/wcdev/app/public/wp-content/plugins/woocommerce/templates/block-notices/notice.php
/Users/jamesallan/local-dev/wcdev/app/public/wp-content/plugins/woocommerce/templates/block-notices/success.php


### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customize-store
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
product-custom-fields
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page
launch-your-store

Disabled Features: experimental-blocks
minified-js
pattern-toolkit-full-composability
product-pre-publish-modal
printful
settings
async-product-editor-category-field
product-editor-template-system

Daily Cron: ✔ Next scheduled: 2024-08-01 02:42:59 +10:00
Options: ✔
Notes: 141
Onboarding: completed

### Action Scheduler ###

Canceled: 7
Oldest: 2024-07-17 05:08:54 +0000
Newest: 2024-07-29 08:16:02 +0000

Complete: 3,514
Oldest: 2024-07-01 00:10:13 +0000
Newest: 2024-07-31 07:09:17 +0000

Failed: 2,662
Oldest: 2024-04-12 09:44:00 +0000
Newest: -0001-11-30 00:00:00 +0000

Pending: 316
Oldest: 2024-07-31 07:56:14 +0000
Newest: 2027-03-18 02:54:20 +0000


### Status report information ###

Generated at: 2024-07-31 17:56:14 +10:00
```
</details>

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/products/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by wavvves** [comment]
Hi @james-allan thank you for reporting this.
I followed your steps, and still, adding to Cart notices were displayed, both for regular products and subscriptions. Can you please quickly check on another env or update the testing steps?

Normal product:


![Image](https://github.com/user-attachments/assets/3688f3d2-f94e-42e1-b082-52efe9abe178)


Subscription:



![Image](https://github.com/user-attachments/assets/7a0920bd-85e3-492e-8c85-fe84eacb7dea)

Furthermore, I tried using wc_add_notice() and it successfully displayed it on Cart and Checkout. I believe we'll need your help to pinpoint the cause for this 🤔



![Image](https://github.com/user-attachments/assets/34a7b467-927c-402d-911d-a05bb5d75138)

**Comment by mikejolley** [comment]
As this hasn't had any updates for 2 weeks I'll mark it closed and unable to reproduce. Thanks.

**Comment by james-allan** [comment]
hmm idk. I thought there might be an issue with my cart page but I've just recreated my cart pages via `WC_Install::create_pages()`. I also made sure to disabled all plugins and enabled a default theme and I could still replicate it today. 

https://github.com/user-attachments/assets/228dc170-96f5-445e-b9cb-ccbb3bcafa49

You'll see in the video that when I add the product to the cart, the cart page doesn't display the notice. When I navigate to the my account page, the queued notice is shown. 

I remember when I filed this issue I searched for the `wc_print_notices()`, I only found a handful of places where that's called. There is a `woocommerce_output_all_notices()` function too and it's hooked onto a bunch of hooks [here](https://github.com/woocommerce/woocommerce/blob/9.2.3/plugins/woocommerce/includes/wc-template-hooks.php#L314) and there is a `woocommerce_before_cart` hook, however it's my understanding that's only fired on the shortcode cart, not the block cart page. 

So Idk, I'd be curious to know if you had a stacktrace for where your `wc_print_notices()` is being called from on the cart page, because I'm not sure what I'm missing.

**Comment by mikejolley** [comment]
@james-allan Appearance > Templates > WooCommerce. Reset your "Page: Cart" template. This contains the notices block which I assume you're missing. Recreating pages will not restore this template.

**Comment by mtnorthrop** [comment]
@james-allan  I ran into this issue as well, but was able to solve it by manually adding the "Store Notices" block to my Cart page, above the "Cart" block. For some reason, the "Store Notices" block was _not_ included on my Cart page when I first installed WooCommerce and generated the store templates. Once I added the "Store Notices" block to my cart page, notifications appeared above the cart, as expected.
